### PR TITLE
EE-1082: Round seigniorage rate in chainspec

### PR DESCRIPTION
**Pull request by mpapierski**
Ref: https://casperlabs.atlassian.net/browse/EE-1082

- Loads round seigniorage rate from chainspec and keeps it in the global state
- Supports upgrading with related tests which also verifies that after upgrading new distribution rate is applied to amounts

Keep in mind there's also change to the default round seigniorage rate which I believe was incorrect in `master`: 

Given parameters in the old comment:

```rust
/// Returns the round seigniorage rate
///
/// Annual issuance: 2%
/// Minimum round exponent: 14
/// Ticks per year: 31536000000
///
/// ```python
/// >>> ((1+0.02)**(2**14/31536000000)-1)
/// 1.0288123020174567e-08
/// >>> 102881230202/10000000000000000000
/// 1.02881230202e-08
/// ```
#[inline]
pub fn round_seigniorage_rate() -> Ratio<U512> {
    const ROUND_SEIGNIORAGE_RATE_NUMER: u64 = 102881230202;
    const ROUND_SEIGNIORAGE_RATE_DENOM: u64 = 10_000_000_000_000_000_000;
    Ratio::new(
        U512::from(ROUND_SEIGNIORAGE_RATE_NUMER),
        U512::from(ROUND_SEIGNIORAGE_RATE_DENOM),
    )
}
```

After running expression `(1+0.02)^((2^14)/31536000000)-1` through factorization process on wolfram alpha we get the fraction [6414/623437335209](https://www.wolframalpha.com/input/?i=factorize+%281%2B0.02%29%5E%28%282%5E14%29%2F31536000000%29-1) which is smaller by about [-2.54341×10^-20](https://www.wolframalpha.com/input/?i=%281%2B0.02%29%5E%28%282%5E14%29%2F31536000000%29-1+%3C+102881230202%2F10000000000000000000)

### CODE REVIEW

File: execution_engine/src/core/engine_state/mod.rs [hunk 1]
```diff
@@ -37,7 +38,9 @@ use casper_types::{
     },
     bytesrepr::{self, ToBytes},
     contracts::{NamedKeys, ENTRY_POINT_NAME_INSTALL, UPGRADE_ENTRY_POINT_NAME},
-    mint, proof_of_stake, runtime_args,
+    mint,
+    mint::{ARG_ROUND_SEIGNIORAGE_RATE, ROUND_SEIGNIORAGE_RATE_KEY},
```
Comment (EdHastingsCasperLabs) [inline execution_engine/src/core/engine_state/mod.rs:42]: self,

File: execution_engine/src/core/engine_state/mod.rs [hunk 2]
```diff
@@ -795,6 +810,26 @@ where
                 .write(locked_funds_period_key, value);
         }
 
+        if let Some(new_round_seigniorage_rate) = upgrade_config.new_round_seigniorage_rate() {
```
Comment (EdHastingsCasperLabs) [inline execution_engine/src/core/engine_state/mod.rs:813]: The spec should be updated to include this (plus new_auction_delay and new_locked_funds_period immediately above it in the logic).

**Review by EdHastingsCasperLabs — APPROVED** [review]
see nitpicks

**Comment by henrytill** [comment]
> After running expression `(1+0.02)^((2^14)/31536000000)-1` through factorization process on wolfram alpha we get the fraction [6414/623437335209](https://www.wolframalpha.com/input/?i=factorize+%281%2B0.02%29%5E%28%282%5E14%29%2F31536000000%29-1) which is smaller by about [-2.54341×10^-20](https://www.wolframalpha.com/input/?i=%281%2B0.02%29%5E%28%282%5E14%29%2F31536000000%29-1+%3C+102881230202%2F10000000000000000000)

:rofl:

**Review by henrytill — APPROVED** [review]
Approving, though please address @EdHastingsCasperLabs's comments.

While the ticket suggested putting this in protocol data instead of global state, I'm okay with this approach.  I do wonder if eventually we will consider putting these upgradeable constants in protocol data though.  I suspect there is  less overhead reading them there than global state, and I also suspect it reduces some of the code complexity.

File: grpc/test_support/src/internal/mod.rs [hunk 3]
```diff
@@ -36,6 +37,14 @@ pub const AUCTION_INSTALL_CONTRACT: &str = "auction_install.wasm";
 pub const DEFAULT_VALIDATOR_SLOTS: u32 = 5;
 pub const DEFAULT_AUCTION_DELAY: u64 = 3;
 pub const DEFAULT_LOCKED_FUNDS_PERIOD: EraId = 15;
+/// Default round seigniorage rate represented as a fractional number.
+///
+/// Annual issuance: 2%
+/// Minimum round exponent: 14
+/// Ticks per year: 31536000000
```
Comment (goral09) [inline grpc/test_support/src/internal/mod.rs:44]: Is this calculated for the minimum round exponent? B/c number of ticks will keep changing.

File: node/src/components/chainspec_loader/config.rs [hunk 4]
```diff
@@ -26,6 +27,14 @@ const DEFAULT_UPGRADE_INSTALLER_PATH: &str = "upgrade_install.wasm";
...
 const DEFAULT_AUCTION_DELAY: u64 = 3;
 const DEFAULT_LOCKED_FUNDS_PERIOD: EraId = 15;
+/// Round seigniorage rate represented as a fractional number
+///
+/// Annual issuance: 2%
+/// Minimum round exponent: 14
+/// Ticks per year: 31536000000
+///
+/// (1+0.02)^((2^14)/31536000000)-1 is expressed as a fraction below.
+const DEFAULT_ROUND_SEIGNIORAGE_RATE: Ratio<u64> = Ratio::new_raw(6414, 623437335209);
```
Comment (afck) [inline node/src/components/chainspec_loader/config.rs:37]: Should this have a default value at all?
I don't think we have a default round exponent.

File: resources/production/chainspec.toml [hunk 5]
```diff
@@ -29,6 +29,14 @@ auction_delay = 3
...
 # (other than the last one), they are paid out in the last block of era N + locked_funds_period.
 locked_funds_period = 15
+# Round seigniorage rate represented as a fractional number
+#
+# Annual issuance: 2%
+# Minimum round exponent: 14
+# Ticks per year: 31536000000
+#
+# (1+0.02)^((2^14)/31536000000)-1 is expressed as a fraction below.
+round_seigniorage_rate = [6_414, 623_437_335_209]
```
Comment (afck) [inline resources/production/chainspec.toml:39]: But in this chainspec the minimum round exponent is 17, not 14. Would be good to keep these values in sync.

File: resources/test/valid/chainspec.toml [hunk 6]
```diff
@@ -10,6 +10,7 @@ accounts_path = 'accounts.csv'
 validator_slots = 5
 auction_delay = 3
 locked_funds_period = 0
+round_seigniorage_rate = [6_414, 623_437_335_209]
```
Comment (afck) [inline resources/test/valid/chainspec.toml:13]: …and here it's 13. (I guess it doesn't matter in the tests; but maybe we can just set it to 14?)

File: utils/nctl/templates/chainspec.toml [hunk 7]
```diff
@@ -29,6 +29,15 @@ auction_delay = 3
 # The delay for the payout of funds, in eras. If a withdraw request is included in a block in era N
 # (other than the last one), they are paid out in the last block of era N + locked_funds_period.
 locked_funds_period = 15
+# Round seigniorage rate represented as a fractional number
```
Comment (afck) [inline utils/nctl/templates/chainspec.toml:32]: Maybe a bit clearer: "…as a fraction of the total supply"
(At first I wasn't sure whether it's in motes.)

File: utils/nctl/templates/chainspec.toml [hunk 8]
```diff
@@ -29,6 +29,15 @@ auction_delay = 3
...
+#
+# Annual issuance: 2%
+# Minimum round exponent: 14
+# Ticks per year: 31536000000
+#
+# After factoring (1+0.02)^((2^14)/31536000000)-1 we get a fraction 6414/623437335209
```
Comment (afck) [inline utils/nctl/templates/chainspec.toml:38]: In this file the minimum round exponent is 12, not 14.

File: grpc/test_support/src/internal/mod.rs (hunk 3 above)
Comment (afck) [reply grpc/test_support/src/internal/mod.rs:44]: It's the number of _milliseconds_ per year.
If the round length is 2<sup>14</sup> milliseconds, there are 31536000000 / 2<sup>14</sup> _rounds_ per year.

**Comment by mpapierski** [comment]
> 
> 
> Approving, though please address @EdHastingsCasperLabs's comments.
> 
> While the ticket suggested putting this in protocol data instead of global state, I'm okay with this approach. I do wonder if eventually we will consider putting these upgradeable constants in protocol data though. I suspect there is less overhead reading them there than global state, and I also suspect it reduces some of the code complexity.

Sorry I unconsciously followed pattern established earlier by passing values to the installer.system contracts

> 
> 
> Approving, though please address @EdHastingsCasperLabs's comments.
> 
> While the ticket suggested putting this in protocol data instead of global state, I'm okay with this approach. I do wonder if eventually we will consider putting these upgradeable constants in protocol data though. I suspect there is less overhead reading them there than global state, and I also suspect it reduces some of the code complexity.

Apologizes I unconsciously followed existing pattern of passing values to the respective installer contracts. I'm all in for putting those in protocol data, although that seems problematic as long as we have to support `use-system-contracts` feature since there's no existing way to get the protocol data from within contract. Perhaps we can revisit this once we'd get rid of system contracts altogether

File: execution_engine/src/core/engine_state/mod.rs (hunk 2 above)
Comment (mpapierski) [reply execution_engine/src/core/engine_state/mod.rs:813]: Did this

File: node/src/components/chainspec_loader/config.rs (hunk 4 above)
Comment (mpapierski) [reply node/src/components/chainspec_loader/config.rs:37]: I see your point though I need _something_ for `impl Default for Genesis` below. I used the default value as it was hardcoded in EE so far

Comment (goral09) [reply node/src/components/chainspec_loader/config.rs:37]: We shouldn't have `Default for Genesis` at all :grimacing:

File: execution_engine/src/core/engine_state/mod.rs (hunk 1 above)
Comment (mpapierski) [reply execution_engine/src/core/engine_state/mod.rs]: Did this

File: resources/local/chainspec.toml [hunk 9]
```diff
@@ -29,6 +29,14 @@ auction_delay = 3
 # The delay for the payout of funds, in eras. If a withdraw request is included in a block in era N
 # (other than the last one), they are paid out in the last block of era N + locked_funds_period.
 locked_funds_period = 15
+# Round seigniorage rate represented as a fractional number
```
Comment (afck) [inline resources/local/chainspec.toml:32]: Sorry, I actually meant the first line of the comment:
```suggestion
# Round seigniorage rate represented as a fraction of the total supply
```

(It's a bit redundant to say something similar in the last line below; also in the other files.)

**Review by afck — APPROVED** [review]
I also don't understand why `Genesis` needs a default; but that wasn't introduced in this PR, of course.

**Pull request merged**